Cabal/Distribution/PackageDescription/Configuration.hs

```diff
@@ -351,18 +353,18 @@ overallDependencies enabled (TargetSet targets) = mconcat depss
 -- | Collect up the targets in a TargetSet of tagged targets, storing the
 -- dependencies as we go.
 flattenTaggedTargets :: TargetSet PDTagged -> (Maybe Library, [(UnqualComponentName, Component)])
-flattenTaggedTargets (TargetSet targets) = foldr untag (Nothing, []) targets
-  where
-    untag (_, Lib _) (Just _, _) = userBug "Only one library expected"
-    untag (_, Lib l) (Nothing, comps) = (Just l, comps)
-    untag (_, SubComp n c) (mb_lib, comps)
-        | any ((== n) . fst) comps =
-          userBug $ "There exist several components with the same name: '" ++ unUnqualComponentName n ++ "'"
-
-        | otherwise = (mb_lib, (n, c) : comps)
-
-    untag (_, PDNull) x = x  -- actually this should not happen, but let's be liberal
-
+flattenTaggedTargets (TargetSet targets) = foldr untag (Nothing, []) targets where
+  untag (depMap, pdTagged) accum = case (pdTagged, accum) of
+    (Lib _, (Just _, _)) -> userBug "Only one library expected"
+    (Lib l, (Nothing, comps)) -> (Just $ redoBD l, comps)
+    (SubComp n c, (mb_lib, comps))
+      | any ((== n) . fst) comps ->
+        userBug $ "There exist several components with the same name: '" ++ display n ++ "'"
+      | otherwise -> (mb_lib, (n, redoBD c) : comps)
+    (PDNull, x) -> x  -- actually this should not happen, but let's be liberal
+    where
+      redoBD :: L.HasBuildInfo a => a -> a
+      redoBD = set L.targetBuildDepends $ fromDepMap depMap
```

Comment:
What does this do?

Reply:
It's been a while, but pretty sure this is the problematic bit via your choice a year ago. The difference between the new version and the old version is now we updated the components' `targetBuildDepends` with the `depMaps` of each target, whereas before those just updated the top-level build depends at https://github.com/haskell/cabal/pull/4383/files#diff-12c8f4f2a1b901f8cbc37a619afafcb7L450.